pkg/events/types.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package events
+
+import (
+	"fmt"
+	"sync"
+
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+var (
+	registeredEventTypes = map[string]*EventType{}
+	mu                   sync.Mutex
+)
+
+// EventType describes the type of an event.
+type EventType struct {
+	name   string
+	source string
+	desc   protoreflect.MessageDescriptor
+}
+
+// Name returns the unique name for the event type.
+func (t EventType) Name() string {
+	return t.name
+}
+
+// Source returns the source for the event type.
+func (t EventType) Source() string {
+	return t.source
+}
+
+// Descriptor returns the message descriptor for messages of the event type.
+func (t EventType) Descriptor() protoreflect.MessageDescriptor {
+	return t.desc
+}
+
+// RegisterType registers a new event type. It is intended for use in init()
+// functions for each event type. It will panic if errors are encountered.
+func RegisterType(name, source string, desc protoreflect.MessageDescriptor) *EventType {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if name == "" {
+		panic("event name must be set")
+	}
+	if source == "" {
+		panic("event source must be set")
+	}
+	if desc == nil {
+		panic("event descriptor must be set")
+	}
+	if _, ok := registeredEventTypes[name]; ok {
+		panic("event has already been registered: " + name)
```

Comment:
Instead of a panic, should this be a warning? Seems low impact to try to register a type twice.

Reply:
If for some reason two callers were registering the same name but different types, it could lead to unexpected runtime errors. I think it is better to panic on start-up initialization in this case, as this is intended to be called from `init()` functions.

Reply:
Thanks for the explanation, I'm good with keeping this as is